enhancements/builds/multi-arch-builds.md

```diff
@@ -0,0 +1,456 @@
+* Add an optional target platforms []string to github.com/openshift/api/openshiftcontrolplane/v1.BuildOverridesConfig.
+* Add an optional "archNodeSelector" *bool to github.com/openshift/api/openshiftcontrolplane/v1.BuildOverridesConfig.
+* Add an optional "manifestList" *bool to github.com/openshift/api/openshiftcontrolplane/v1.BuildOverridesConfig.
+* Figure out where the platforms list for builds should come from if the configured BuildOverridesConfig doesn't have one.
+* Teach openshift-controller-manager to launch, monitor, and when necessary cancel the N+1 pods, setting the target platform and output location differently in the encoded Build values it passes to the various platform-specific build pods, unless "manifestList" is explicitly disabled.
+  * Build pods will gain a label which `oc logs` can use to select which of the build's pods it will read logs from.
+  * If a node with the corresponding OS and architecture selector labels (e.g. "linux", "amd64") is not able to build for any requested variant (e.g. "", "v2", etc.), more refinement will be required.
+  * Unless BuildOverridesConfig.archNodeSelector is explicitly disabled, each pod will have a "kubernetes.io/arch" value added to its "nodeSelector".
+* Update `oc logs` updated Build and BuildConfig types. Update log viewing UX.
+* Update `oc describe` for updated Build and BuildConfig types. Update build progress / status UX.
+* Extend `oc new-app` to check the architectures of the images that the BuildConfig it's about to generate uses.
+
+### Risks and Mitigations
+
+* This proposal doesn't grant new privileges to any of the components involved, but it should still get a security review.
+* Each pod will generate its own logs, so this will require updates to the UX to present them distinctly.
+* Registries which don't allow untagging: ignore error results; we're only untagging to try to be nice.
+* While we can dutifully mark a built image as e.g. linux/amd64 or linux/amd64/v2 when we're told to build one or the other, we're dependent on the build's Dockerfile and the commands it runs to actually produce binaries that are correct for the target platform variant.
+
+### Drawbacks
+
+* Limited to using architectures available on the cluster.
+* Builds which use base images which are not available for all of the cluster's architectures will require manual configuration, either of the build controller to limit the target platforms for all builds, or of individual affected Builds or BuildConfigs to limit the target platforms for them.
+* CI for multiple architectures adds failure points.
+* CI for multiple architectures will necessarily cost more.
+* CI for multiple architectures will necessarily take longer to complete.
+
+## Design Details
+
+### Open Questions [optional]
+
+* If the openshift-controller-manager configuration doesn't include a list of target platforms, from where in the multi-arch cluster should that information be sourced?
+
+### Test Plan
```

Comment:
we should also mention here that our current CI jobs need to be adapted to use this new builder. does anything special need to be done for that? i think it should just work on a multi-arch cluster once we have this in place

Reply:
if you actually expect to produce manifestlist content on our CI system, there are definitely changes that will be needed.

first and foremost, the CI system would need to produce buildconfigs that specify the "manifestlist=true" bool or whatever.

but that's only scratching the surface of the orchestration logic that probably would require updating.

if by "just work" you mean the existing flows (producing single arch content) should just work, yes, that should just keep working mostly...   though if you drop a single arch build on a multiarch cluster, you need to have properly nodeselectors on your buildconfig to ensure you build arch you expect, on the node type that is required.  So at a minimum that would be needed.